Implement W3C tracestate limitations

andriikushch opened the pull request:
Implement https://www.w3.org/TR/trace-context/#tracestate-header

andrewslotin commented on w3ctrace/state.go:
```diff
@@ -19,13 +19,34 @@ type State []string
-		if st == "" {
-			continue
+	// comasLimit is used to prevent handling of malformed values that contain a lot of comas
+	const comasLimit = 256
+	rawEntries := strings.SplitN(traceStateValue, ",", comasLimit)
```
```suggestion
	const maxEntries = 256
	rawEntries := strings.SplitN(traceStateValue, ",", maxEntries)
```

andriikushch replied:
This number is something just bigger than a reasonable amount of tracestate entries (in raw header). To prevent a situation when malicious user sent values like ",,,,,,,,,,,,,,,,,,,,,,,,,(and many comas after)" and system, will waste a CPU resource. If you think it is too much, we can use `Split` instead of `SplitN`. What do you think?

andrewslotin commented on w3ctrace/state.go:
```diff
@@ -19,13 +19,34 @@ type State []string
-		if st == "" {
-			continue
+	// comasLimit is used to prevent handling of malformed values that contain a lot of comas
+	const comasLimit = 256
```
Where does this number come from?

andriikushch replied:
This number is something just bigger than a reasonable amount of tracestate entries (in raw header). To prevent a situation when malicious user sent values like ",,,,,,,,,,,,,,,,,,,,,,,,,(and many comas after)" and system, will waste a CPU resource. If you think it is too much, we can use Split instead of SplitN. What do you think?

andrewslotin replied:
Given that the size of HTTP headers section is limited, I'd say it's unlikely this will become a problem.

andriikushch replied:
Removed

andrewslotin commented on w3ctrace/state.go:
```diff
@@ -19,13 +19,34 @@ type State []string
+				break
+			}
+
+			st = strings.TrimSpace(st)
```
This would potentially modify list members now owned by Instana, which may break trace correlation downstream. I'd suggest we don't alter the value that we did not add to the state.

andriikushch replied:
I tried to keep the same behavior as before https://github.com/instana/go-sensor/blob/master/w3ctrace/state.go#L23.

andrewslotin replied:
This does not look correct either 😅

andriikushch replied:
:) Let's change it then.

andrewslotin replied:
Yes, but according to the spec the `value` may include whitespace characters (`%x20`) and since they are indistinguishable from optional whitespace (`OWS`) in the end, I'd suggest we don't modify the elements that we don't add.

andrewslotin commented on w3ctrace/state.go:
```diff
@@ -19,13 +19,34 @@ type State []string
+		}
+	} else {
+		for _, st := range rawEntries {
+			st = strings.TrimSpace(st)
```
Same as above, this would alter a value that we don't own.

andriikushch replied:
Answered in the previous comment.

andrewslotin reviewed (changes requested):
I'm not entirely sure the truncation logic follows the spec. At the moment for the `traceparent` header containing more than 32 key-value pairs these changes:
* Starting from beginning any entry longer than 128 characters or empty, not counting leading/trailing spaces is removed

This approach does not guarantee that resulting set would contain 32 elements or less, e.g. if state does not contain any empty/large key-value pairs.

On the other hand the W3C prescribes the following algorithm:
1. The entries at the end of `tracestate` SHOULD be removed first
2. First step is to remove entries longer then 128 characters
3. If the previous step did not produce a list of 32 elements or less, truncate the last `len - 32` elements.

Let me know if I'm missing something

andriikushch commented:
> I'm not entirely sure the truncation logic follows the spec. At the moment for the `traceparent` header containing more than 32 key-value pairs these changes:
> 
> * Starting from beginning any entry longer than 128 characters or empty, not counting leading/trailing spaces is removed
> 
> This approach does not guarantee that resulting set would contain 32 elements or less, e.g. if state does not contain any empty/large key-value pairs.

I think this case is covered here https://github.com/instana/go-sensor/pull/287/files#diff-4d31341d37f313670fd6f87a6a018dccc0301c44aab43258fe1f99d8c3341942R34, isn't it? 

> 
> On the other hand the W3C prescribes the following algorithm:
> 
> 1. The entries at the end of `tracestate` SHOULD be removed first

It looks like this is missed here: `The entries at the end of tracestate SHOULD be removed first`, but I can't find it in a spec. Can you please share a link to this line, because I found only two next statements?
> 2. First step is to remove entries longer then 128 characters
> 3. If the previous step did not produce a list of 32 elements or less, truncate the last `len - 32` elements.
> 
> Let me know if I'm missing something

andrewslotin commented on w3ctrace/state_test.go:
```diff
@@ -30,6 +31,18 @@ func TestParseState(t *testing.T) {
 			Expected: w3ctrace.State{"rojo=00f067aa0ba902b7", "congo=t61rcWkgMzE"},
 		},
+		"with 33 list items": {
+			Header:   strings.TrimLeft(strings.Repeat("rojo=00f067aa0ba902b7,", 33), ","),
```
```suggestion
			Header:   strings.TrimRight(strings.Repeat("rojo=00f067aa0ba902b7,", 33), ","),
```
Otherwise it's leaving the trailing comma at the end, which is not an issue for the test, but contradicts the test name

andriikushch replied:
Good catch - typo 👍

andrewslotin commented on w3ctrace/state_test.go:
```diff
@@ -30,6 +31,18 @@ func TestParseState(t *testing.T) {
+			Expected: strings.Split(strings.TrimRight(strings.Repeat("rojo=00f067aa0ba902b7,", 32), ","), ","),
+		},
+		"with 34 list items, with long one at the beginning": {
+			Header:   "rojo=" + strings.Repeat("a", 129) + "," + strings.TrimLeft(strings.Repeat("rojo=00f067aa0ba902b7,", 33), ","),
```
```suggestion
			Header:   "rojo=" + strings.Repeat("a", 129) + "," + strings.TrimRight(strings.Repeat("rojo=00f067aa0ba902b7,", 33), ","),
```
Same as above

andrewslotin commented:
> I think this case is covered here https://github.com/instana/go-sensor/pull/287/files#diff-4d31341d37f313670fd6f87a6a018dccc0301c44aab43258fe1f99d8c3341942R34, isn't it?

Right, I was confused by the condition that breaks the loop 👍 

> It looks like this is missed here: "The entries at the end of tracestate SHOULD be removed first", but I can't find it in a spec. Can you please share a link to this line, because I found only two next statements?

Here is the quote: 

> In a situation where tracestate needs to be truncated due to size limitations, the vendor MUST truncate whole entries. Entries larger than 128 characters long SHOULD be removed first. Then entries SHOULD be removed starting from the end of `tracestate`. 

Same as above, I believe I misread the loop condition. I believe the part `rawEntriesLen-filtered <= 32` is redundant, at least removing it does not fail any of your tests.

andrewslotin commented on w3ctrace/state.go:
```diff
@@ -19,13 +19,29 @@ type State []string
+		}
+	} else {
+		for _, st := range rawEntries {
+			if st == "" {
+				continue
+			}
```
As a suggestion, if we stop filtering empty list items for each request, we could spare some CPU cycles and use a guard-clause here instead of `else`

andriikushch replied:
Like this https://github.com/instana/go-sensor/pull/287/files#diff-8e3fa4ec5050520f1e8157ac884ce463ca2553109301a7e640e2e0a5f9927aa7R38?

andriikushch commented:
> Same as above, I believe I misread the loop condition. I believe the part `rawEntriesLen-filtered <= 32` is redundant, at least removing it does not fail any of your tests.

I think you are right.

andriikushch commented:
@andrewslotin I changed the logic and added a few more tests. I hope now the code looks more clear and straightforward. Please take a look when you will have time.

andrewslotin commented on w3ctrace/state.go:
```diff
@@ -11,24 +11,56 @@ import (
-		if st == "" {
-			continue
+	entries := strings.Split(traceStateValue, ",")
+	if len(entries) == 1 && entries[0] == "" {
+		return state, nil
+	}
```
This looks like a premature optimization: the `range` statement inside `filterEmptyItems` will not be executed for an empty list, so the check at L35 will suffice.

andriikushch replied:
This is changed now 👍

andrewslotin commented on w3ctrace/state.go:
```diff
@@ -11,24 +11,56 @@ import (
+	filteredEntries := filterEmptyItems(entries)
+	filteredEntriesLen := len(filteredEntries)
+
+	if filteredEntriesLen == 0 {
+		return state, nil
+	}
```
Actually this check is also redundant, since the following `if` statement condition will result to `false` in case the list is empty.

andriikushch replied:
This is changed now 👍

andrewslotin commented on w3ctrace/state.go:
```diff
@@ -11,24 +11,56 @@ import (
+		return state, nil
+	}
+
+	if filteredEntriesLen > maxKVPairs {
```
Let's change it to an early return to reduce the cognitive complexity.

andriikushch replied:
👍

andrewslotin commented on w3ctrace/state.go:
```diff
@@ -109,3 +141,16 @@ func (st State) String() string {
 }
+
+func filterEmptyItems(entries []string) []string {
+	var entriesWithoutEmptyItems []string
```
As a suggestion, you could consider preallocating the slice here, since its max capacity is known.

andriikushch replied:
This part now has a different implementation, without extra memory allocation.

andrewslotin commented on w3ctrace/state.go:
```diff
@@ -11,24 +11,56 @@ import (
+		filtered := 0
+
+		for k, st := range filteredEntries {
+			if len(state) == maxKVPairs {
+				break
+			}
+
+			// check if enough elements were filtered already
+			if filteredEntriesLen-filtered <= maxKVPairs {
+				return append(state, filteredEntries[k:]...), nil
+			}
```
I still find it hard to understand the logic here, since one needs to consider the size two lists at once. I believe we can do this with one list, removing the items longer than 128 chars until either it contains 32 elements, or there are no large elements left. And then return first `min(len(state), 32)` elements of it?

andriikushch replied:
Thank you for the idea. I have changed the implementation and added a new test case.

1. I removed extra memory allocations for filtering.
2. I changed the filtering function, now I think it looks close to what you propose.

Please check when you will have time.

andrewslotin commented on w3ctrace/state.go:
```diff
@@ -11,24 +11,40 @@ import (
+	filteredStates = append(filteredStates, states[i:]...)
 
-	return state, nil
+	return filteredStates[:min(len(filteredStates), maxKVPairs)], nil
```
Again just for readability reasons let's inline the `min()` function:
```suggestion
	if len(filteredStates) > maxKVPairs {
		filteredStates = filteredStates[:maxKVPairs]
	}
	
	return filteredStates
```

andriikushch replied:
Added

andrewslotin reviewed (approved):
Overall I think it looks good 👍 Just a couple of suggestions to improve readability even further.

andriikushch commented on w3ctrace/state.go:
```diff
@@ -11,24 +11,40 @@ import (
 func ParseState(traceStateValue string) (State, error) {
-	var state State
+	states := filterEmptyItems(strings.Split(traceStateValue, ","))
+	statesLen := len(states)
```
done
